sql/moz-fx-data-shared-prod/udf_js/bootstrap_percentile_ci/udf.sql

```diff
@@ -0,0 +1,212 @@
+/*
+Calculates a confidence interval using an efficient bootstrap sampling technique for a
+given percentile of a histogram.
+
+See:
+- https://mozilla-hub.atlassian.net/browse/DS-2128
+- https://engineering.atspotify.com/2022/03/comparing-quantiles-at-scale-in-online-a-b-testing/
+
+Users must specify an array of percentiles of interest as the first parameter and a
+histogram struct as the second.
+*/
+
+CREATE OR REPLACE FUNCTION udf_js.bootstrap_percentile_ci(
+  percentiles ARRAY<INT64>,
+  histogram STRUCT<values ARRAY<STRUCT<key FLOAT64, value FLOAT64>>>,
+  metric STRING
+)
+RETURNS ARRAY<STRUCT<
+    metric STRING,
+    statistic STRING,
+    point FLOAT64,
+    lower FLOAT64,
+    upper FLOAT64,
+    parameter STRING
+>> DETERMINISTIC
+LANGUAGE js OPTIONS ( library=['gs://moz-fx-data-circleci-tests-bigquery-etl/bq_udf_library/qbinom.js'])
+AS
+"""
+  function numericSort(arr) {
+    arr.sort(function(a, b) {return a - b;});
+  }
+
+  function validatePercentile(percentile) {
+    if (percentile < 0 || percentile > 100) {
+      throw "percentile must be a value between 0 and 100";
+    }
+  }
+  
+  function getQuantileSamples(percentile, histogram, n_samples) {
+    validatePercentile(percentile);
+
+    /* expand the histogram */
+    var full_histogram = histogram.map(bin => Array(bin.value).fill(parseFloat(bin.key))).reduce((prev, curr) => prev.concat(curr));
+    numericSort(full_histogram);
```

Comment:
As discussed, this is causing `UDF memory exceeded` errors for certain histograms like `gc_ms`. But it _should_ be safe to remove at least this operation since it seems to be guaranteed that histograms remain sorted.

We should add some kind of check though to throw an error if that is not the case.

Reply:
https://github.com/mozilla/bigquery-etl/pull/3517/commits/79997ec3355412759165b1dd58ed736d2e235813 changes the order of operations so that we first sort the histogram based on the keys, then we build the `full_histogram`. As we talked about offline, sorting the histogram is much faster and the `map` method will process the keys in order.